crypto/rsa/rsa_ossl.c

```diff
@@ -892,22 +892,20 @@ static int rsa_ossl_mod_exp(BIGNUM *r0, const BIGNUM *I, RSA *rsa, BN_CTX *ctx)
         if (!BN_is_zero(vrfy)) {
             /*
              * 'I' and 'vrfy' aren't congruent mod n. Don't leak
-             * miscalculated CRT output, just do a raw (slower) mod_exp and
-             * return that instead.
+             * miscalculated CRT output to prevent Bellcore attack.
+             * Do NOT compute a raw (slower) mod_exp, clear sensitive variables
+             * and return an error to prevent Sébastien CARRE, Adrien FACON, 
+             * Sylvain GUILLEY and Matthieu DESJARDINS fault attack on message.
              */
-
-            BIGNUM *d = BN_new();
-            if (d == NULL)
-                goto err;
-            BN_with_flags(d, rsa->d, BN_FLG_CONSTTIME);
-
-            if (!rsa->meth->bn_mod_exp(r0, I, d, rsa->n, ctx,
-                                       rsa->_method_mod_n)) {
-                BN_free(d);
-                goto err;
-            }
-            /* We MUST free d before any further use of rsa->d */
-            BN_free(d);
+          
+            /* Prevent prime factor leak due to message fault attack. */
+            BN_clear(I);
```

Comment:
You can't clear I, it's marked as const.

Reply:
You're right.  However it seems important to clear it because the fault we found change `I` to `I modulo q` that leaks the prime factor `q`. I guess removing the const flag is not a good option. Maybe copying `I` into another temporary and not constant variable could be used? By this way we can use this new variable instead of `I` all along the code and clearing it is possible.

Reply:
This function should not be modifying `I` at all. The const should prevent that, nor do I see any code that does it. As far as I can see all the other variables created with BN_with_flags() that use `I` also don't modify it.

Reply:
Actually, we study that in the rowhammer attack context in the assembly level. It means that const is not an issue for an attacker. I think copying `I` into another variable and use this new variable makes `I` safe and unchanged. I commit a minor change to show that idea. Perhaps not the better solution for you?

Reply:
Neither `I` nor its copy are modified by this function. In fact the value is typically public anyway. (It's the ciphertext you are decrypting or the padded payload you are signing.) Are you sure you're not trying to clear a different variable? This is making a copy for the sake of clearing it later.

Reply:
I agree with the fact that neither `I` nor its copy are modified by the function. However, the fault consists of using rowhammer attack to change `BN_mod(r1,c,rsa->q,ctx)` to `BN_mod(c,c,rsa->q,ctx)`. Since `c` and `I` are the same (except a flag), it results in a modification of `I` (even with const in C) to be equal to `I mod q`. If `I` is a public variable, then the attacker can retrieve `I mod q` with the fault (that is why I wanted to clear it) and then deduce q by using a gcd. If you think it is not possible, then clearing `I` is indeed not necessary and simply returning an error as BoringSSL does is also good.

Reply:
> rowhammer attack to change `BN_mod(r1,c,rsa->q,ctx)` to `BN_mod(c,c,rsa->q,ctx)`.

OK, but if you can change one pointer to another pointer using rowhammer in a crypto library, can't you do even more devastating attacks? My understanding of this fault attack mitigation is that the original kind of fault was claimed to be so likely that it can even happen accidentally. Something like rowhammer replacing pointers in memory seems like it would be out of scope for what common crypto libraries are trying to defend against.

Reply:
Sure. That is perfectly true. Actually we studied it in a rowhammer context (that can be deterministic according to https://gruss.cc/files/drammer.pdf) but it can be applied in another context: This is a fault attack as well as a Bellcore attack for which OpenSSL has a protection. 

I also agree with the choice of BoringSSL (and NSS) to consider other faults that CRT fault by simply returning an error even if I totally agree that the probability of such attack could be low (until one find a way to improve it).

According to your answers, clearing variables is definitely not a good idea because it can indeed open timing attack issues. I assume those variables are not cleared later for the same reason. Of course this leads the miscalculated, and possibly sensitive, faulted signature to remain in memory. But I guess that if one can read this signature in memory it can also directly read the prime factors.

So finally, I think the best solution is to simply returning an error without trying non-CRT signature as BoringSSL and NSS do. This assume that the bad signature is not given to the one how requests it. 

Before I commit another patch I would want to know your opinion.